[CIT-316] Document the mapper component (#202)
thin-edge/thin-edge.io: docs/src/architecture/mapper.md

# The tedge-mapper

The tedge-mapper is a key concept to support multiple cloud providers.
The purpose is to translate
messages written using the cloud-agnostic [Thin Edge JSON format](thin-edge-json.md),
into cloud-specific messages.

tedge-mapper is composed of multiple cloud-specific mappers, such as Cumulocity mapper and Azure mapper.
Each mapper is responsible for its dedicated cloud.

> > Note: The tedge-mapper contains the Cumulocity mapper only currently.

Reviewer: I would prefer not to mention that explicitly here. We will have it soon.

Reviewer 2: We need to say how to run these different mappers.

These specific mappers are launched by the respective `tedge connect` command.  For instance, `tedge connect c8y` establishes a bridge to Cumulocity and launches a Cumulocity mapper that translates the messages in the background.

> A mapper subscribes the reserved MQTT topic `tedge/measurements` with the QoS level At Least Once.

Reviewer: ```suggestion
A mapper subscribes to the reserved MQTT topic `tedge/measurements` with the QoS level 1 (At least once).
```

Reviewer: Using level 1 instead of full name makes reading more fluent.

> The messages that arrived in the mapper should be formed in the [Thin Edge JSON](thin-edge-json.md) format. 

Reviewer: ```suggestion
The messages that arrive in the mapper should be formed in the [Thin Edge JSON](thin-edge-json.md) format. 
```

The mapper verifies whether the arrived messages are correctly formatted,
in case the verification fails, the mapper publishes a corresponded error message

> on the topic `tedge/errors` with the QoS level At Least Once.

Reviewer: ```suggestion
on the topic `tedge/errors` with the QoS level 1.
```

Reviewer 2: I would make explicit both the QoS level and the QoS semantics, as in "QoS level 1 (at least once)".

Author: Changed to QoS level 1 (at least once)

Here is an example if you publish invalid Thin Edge JSON messages on `tedge/measurements`:

```shell
$ tedge mqtt pub tedge/measurements '{"temperature": 23,"pressure": 220'
$ tedge mqtt pub tedge/measurements '{"temperature": 23,"time": 220}'
```

Then, you'll receive error messages from the mapper on the topic `tedge/errors`:

```shell
$ ./tedge mqtt sub tedge/errors
[tedge/errors] Invalid JSON: Unexpected end of JSON: {"temperature":23,"pressure":220
[tedge/errors] Not a timestamp: the time value must be an ISO8601 timestamp string in the YYYY-MM-DDThh:mm:ss.sss.±hh:mm format, not a number.
```

> Once the mapper receives a correctly formatted message, 

Reviewer 2: When the mapper receives ...

the message will be translated into a cloud-specific format.

## Cumulocity mapper
The Cumulocity mapper translates [Thin Edge JSON](thin-edge-json.md) into Cumulocity's [JSON via MQTT](https://cumulocity.com/guides/device-sdk/mqtt/#json).

> Translated messages will be published in the topic `c8y/measurement/measurements/create`.

Reviewer 2: ```suggestion
The translated messages are published on the topic `c8y/measurement/measurements/create` from where they are forwarded to Cumulocity.
```

Example in Thin Edge JSON:

```json
{
	"temperature": 23
}
```

Translated into JSON via MQTT by the Cumulocity mapper:

```json
{
	"type": "ThinEdgeMeasurement",
	"time": "2021-04-22T17:05:26.958340390+00:00",
	"temperature": {
		"temperature": {
			"value": 23
		}
	}
}
```

You can see the Cumulocity mapper added the three things which are not defined before translation.
1. `type` is added.
2. `time` is added.

> 3. Another `temperature` is added.

Reviewer: ```suggestion
3. Another hierarchy level is added, as required by the cumulocity data model. String `temperature` is used as fragment and series.
```
Previous wording sounded like it's another temp measurement that is added.

Reviewer: Ok i now read also the sentence below. But still its better to have it more explicit maybe and not misleading.

Reviewer 2: I agree with Sebastian suggestion.

Going through 1), the `type` is a mandatory field in the Cumulocity's JSON via MQTT manner,
therefore, the Cumulocity mapper always adds `ThinEdgeMeasurement` as a type.
This value is not configurable by users.

Next, 2) `time` will be added by the mapper **only when it is not specified in a received Thin Edge JSON message**.

> In this case, the timezone is always UTC+0. If you want another timezone, specify the time filed in Thin Edge JSON.

Reviewer 2: This is incorrect. With the latest version of the Cumulocity translator, the time zone is the device's local time zone.

Author: Oh yes. I tested in the debian package version. The latest version uses my local timezone.

# The tedge-mapper

The tedge-mapper is a key concept to support multiple cloud providers.
The purpose is to translate
messages written using the cloud-agnostic [Thin Edge JSON format](thin-edge-json.md),
into cloud-specific messages.

The tedge-mapper is composed of multiple cloud-specific mappers, such as Cumulocity mapper and Azure mapper.
Each mapper is responsible for its dedicated cloud.
These specific mappers are launched by the respective `tedge connect` command.
For instance, `tedge connect c8y` establishes a bridge to Cumulocity and launches a Cumulocity mapper
that translates the messages in the background.

A mapper subscribes to the reserved MQTT topic `tedge/measurements` with the QoS level 1 (at least once).

> The messages that arrive in the mapper should be formed in the [Thin Edge JSON](thin-edge-json.md) format.

Reviewer: We should change the use of Thin Edge JSON consistently to thin-edge.io JSON. But Ok for now. We need to change that together in all docs then.
